packages/react-dom/src/events/DOMEventResponderSystem.js

```diff
@@ -0,0 +1,114 @@
+// To ensure we fire the right responders for given events, we check
+// if the incoming event type is actually relevant for an event
+// responder. Instead of doing an O(n) lookup on the event responder
+// target event types array each time, we instead create a Set for
+// faster O(1) lookups.
+export const eventResponderValidEventTypes: Map<
+  ReactEventResponder,
+  Set<DOMTopLevelEventType>,
+> = new Map();
+
+function createValidEventTypeSet(targetEventTypes): Set<DOMTopLevelEventType> {
+  const eventTypeSet = new Set();
+  // Go through each target event type of the event responder
+  for (let i = 0, length = targetEventTypes.length; i < length; ++i) {
+    const targetEventType = targetEventTypes[i];
+
+    if (typeof targetEventType === 'string') {
+      eventTypeSet.add(((targetEventType: any): DOMTopLevelEventType));
+    } else {
+      if (__DEV__) {
+        warning(
+          typeof targetEventType === 'object' && targetEventType !== null,
+          'Event Responder: invalid entry in targetEventTypes array. ' +
+            'Entry must be string or an object. Instead, got %s.',
+          targetEventType,
+        );
+      }
+      const targetEventConfigObject = ((targetEventType: any): {
+        name: DOMTopLevelEventType,
+        passive?: boolean,
+        capture?: boolean,
+      });
+      eventTypeSet.add(targetEventConfigObject.name);
+    }
+  }
+  return eventTypeSet;
+}
+
+function handleTopLevelType(
+  topLevelType: DOMTopLevelEventType,
+  fiber: Fiber,
+  context: Object,
+): void {
+  const responder: ReactEventResponder = fiber.type.responder;
+  const props = fiber.memoizedProps;
+  const stateNode = fiber.stateNode;
+  let validEventTypesForResponder = eventResponderValidEventTypes.get(
+    responder,
+  );
+
+  if (validEventTypesForResponder === undefined) {
+    validEventTypesForResponder = createValidEventTypeSet(
+      responder.targetEventTypes,
+    );
+    eventResponderValidEventTypes.set(responder, validEventTypesForResponder);
+  }
+  if (!validEventTypesForResponder.has(topLevelType)) {
+    return;
+  }
+  let state = stateNode.get(responder);
+  if (state === undefined && responder.createInitialState !== undefined) {
+    state = responder.createInitialState(props);
+    stateNode.set(responder, state);
+  }
+  // TODO provide all the props for handleEvent
+  responder.handleEvent(context, props, state);
+}
+
+export function runResponderEventsInBatch(
+  topLevelType: DOMTopLevelEventType,
+  targetFiber: Fiber,
+  nativeEvent: AnyNativeEvent,
+  nativeEventTarget: EventTarget,
+  eventSystemFlags: EventSystemFlags,
+): void {
+  // TODO add proper event context
+  let context = ({}: any);
+  let node = targetFiber;
+  // Traverse up the fiber tree till we find event component fibers.
+  while (node !== null) {
+    if (node.tag === EventComponent) {
+      if (node.alternate !== null && !currentEventComponentFibers.has(node)) {
+        node = node.alternate;
```

Comment:
What's the logic here? How do you choose between current and alternate and why? Can you write the high level algorithm as a comment in more detail?

Reply:
I've updated the code and added a very detailed explanation.